### PR TITLE
Json schema draft v4 support

### PR DESCRIPTION
**Pull request by netmilk**

### CODE REVIEW

File: scripts/build [hunk 1]
```diff
@@ -8,5 +8,6 @@ if [ -d "src/" ]; then
 	#cp -a -R src/* lib/
 fi
 cp test/cucumber/features/node_js/Readme.md ./
-cp src/meta-schema.json lib/meta-schema.json
+cp src/meta-schema-v3.json lib/meta-schema-v3.json
+cp src/meta-schema-v4.json lib/meta-schema-v4.json
```
Comment (kuba-kubula) [inline scripts/build]: I'd prefer to use a `.coffee` file extension so the build script doesn't need to handle it by itself (coffee compile will handle it when building files). But YMMV.

File: src/mixins/validatable-http-message.coffee [hunk 2]
```diff
@@ -14,11 +16,11 @@ class Validatable
   #@return [Object] :headers {ValidationErrors}, :body {ValidationErrors}, :statusCode [Boolean]
   validate: () ->
     @validation = {}
-
+    @validation.version = '2'
```
Comment (kuba-kubula) [inline src/mixins/validatable-http-message.coffee]: :+1:

File: src/mixins/validatable-http-message.coffee [hunk 3]
```diff
@@ -14,11 +16,11 @@ class Validatable
...
     @lowercaseHeaders()
 
-    @validateHeaders() unless @headers == undefined
-    @validateBody() unless @body == undefined
+    @validateHeaders() unless @headers == undefined || @expected.headers == undefined
```
Comment (kuba-kubula) [inline src/mixins/validatable-http-message.coffee]: Why this change? It just brings another type of behavior. Do we really want "not to validate headers" if one of real/expected `headers` is missing?

dtto for `body`

File: src/mixins/validatable-http-message.coffee [hunk 4]
```diff
@@ -14,11 +16,11 @@ class Validatable
...
+    @validateBody() unless @body == undefined || @expected.body == undefined
     @validateStatusCode() unless @statusCode == undefined
```
Comment (kuba-kubula) [inline src/mixins/validatable-http-message.coffee]: Why isnt' the same "or" as above used here too? (check for `expected.statusCode?`)

File: src/validators/json-schema.coffee [hunk 5]
```diff
@@ -90,14 +127,25 @@ class JsonSchema
...
       indexes = [0..data.length - 1]
       indexes.forEach (index) ->
         item = data[index]
-        pointer = item['property'] or []
+
+        if item['property'] == undefined
+          pathArray = []
+        else if item['property'] == null
+          pathArray = []
+        else if item['property'] == [null]
```
Comment (kuba-kubula) [inline src/validators/json-schema.coffee]: You cannot do this type of comparison, it won't work in JS.
Please do `Array.isArray(item['property']) and item['property'].length is 1 and item['property'][0] is null`

Or simply we can remove these checks, they don't seem useable to me.
Can you please provide some way they can actually happen?

File: src/validators/json-schema.coffee [hunk 6]
```diff
@@ -90,14 +127,25 @@ class JsonSchema
...
 
-    #amanda to gavel converter
+    #TV4 to gavel converter
     if data.length > 0
       indexes = [0..data.length - 1]
       indexes.forEach (index) ->
         item = data[index]
-        pointer = item['property'] or []
+
+        if item['property'] == undefined
```
Comment (kuba-kubula) [inline src/validators/json-schema.coffee]: This `if and the next `else if` can be written as `if not item['property]? then pathArray = []`

**Comment by kuba-kubula** [comment]
Great thanks @marcelogo for the great work you've done to bring the basic `tv4`-based schema validations into _gavel.js_. I really appreciate the ground work.

**Comment by yakovkhalinsky** [comment]
@marcelogo :+1:
